src/resolvelib/structs.py

```diff
@@ -62,3 +67,32 @@ def iter_children(self, key):
 
     def iter_parents(self, key):
         return iter(self._backwards[key])
+
+
+class Candidates(object):
+    """Wrap the candidate iterator factory returned by ``find_matches()``.
+
+    Calling `iter()` on this class would invoke the underlying iterator
+    factory, making it a "collection with ordering" that can be iterated
+    through multiple times, but lacks random access methods presented in
+    built-in Python sequence types.
+    """
+
+    def __init__(self, matches):
+        if callable(matches):
+            self._factory = matches
+        elif not isinstance(matches, collections_abc.Sequence):
+            self._factory = functools.partial(iter, list(matches))
+        else:
+            self._factory = functools.partial(iter, matches)
+
+    def __iter__(self):
+        return self._factory()
+
+    def excluding(self, candidate):
+        """Create a new ``Candidates`` instance excluding ``candidate``."""
+
+        def factory():
+            return (c for c in self._factory() if c != candidate)
+
+        return type(self)(factory)
```

Comment:
Nitpick: I think just `return Candidates(c for c in self._factory() if c != candidate)` would fit the bill and is more readable.

Reply:
The extra function is needed, otherwise the iterator gets resolved eagerly and breaks the optimisation.

Reply:
Fix me if I'm wrong, but it seems that if we feed `__init__` an iterator like so, then `self._factory()` would be `iter(matches)`.

Reply:
Correct, but that’s no reusable because `iter(iterator) is iterator`. To make what you want work we need to invent an iterable class that caches the iterator’s content on iteration and return those cached values instead of the original iterator later on. That’s totally doable, but I didn’t like what I came up with and decided to not so that.

Reply:
Ah, I see.